### DISCUSSION TITLE
ch1/ch1-04

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#108] [General]
# ch1/ch1-04

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch1/ch1-04.html

**Comment by SimonJMD** [comment]
这里直接复制代码输出
```
$ go build gopl.io/ch1/lissajous
$ ./lissajous >out.gif
```
生成的gif是无法打开的
但是如果这样调用是可以的
```
	create, _ := os.Create("out.gif")
	lissajous(create)
```
64位win10，go1.19rc2亲测，为什么会这样?

**Reply by chai2010** [reply]
diff看看，应该是windows搞的鬼，把stdout特化了

**Reply by SimonJMD** [reply]
不能说很像，只能说完全不一样了
正常输出文件头部是GIF89a
stdout搞出来的是G I F 8 9 A，每一个都多了一个00，后面更是插入了无数个空格，最后正常生成的gif只有213kB，stdout生成的是314kB

**Reply by SimonJMD** [reply]
我之前是用的IDEA的命令行有这个问题，现在用CMD没有这个问题了，但是CMD出来的只有76kB，然后出来的图像是竖着的-。-

**Reply by SharkAXZ** [reply]
Windows下powershell的问题
关于这个情况，比较明确的解释：
https://stackoverflow.com/questions/56323293/go-generated-animated-gifs-didnt-work-in-windows
https://github.com/golang/go/issues/13746

reference:
https://github.com/golang/go/issues/42337
https://www.cnblogs.com/linxiaoxu/p/16187331.html

**Reply by fool-yf** [reply]
mac m3 pro 也必须在build之后，运行./lissajous >out.gif命令才能正常展示

**Comment by Zengyinjun** [comment]
```// Lissajous generates GIF animations of random Lissajous figures.
package main

import (
	"image"
	"image/color"
	"image/gif"
	"io"
	"math"
	"math/rand"
	"os"
	"time"
)

var palette = []color.Color{color.White, color.RGBA{0, 0xff, 0, 0xff},
	color.RGBA{0, 0, 0xff, 0xff}, color.RGBA{0xff, 0, 0, 0xff},
	color.RGBA{0x22, 0x22, 0x22, 0xff}, color.RGBA{0xff, 0x44, 0x15, 0xff},
	color.RGBA{0x44, 0x62, 0x12, 0xff}, color.RGBA{0xff, 0xff, 0x15, 0xff}}

const (
	whiteIndex = 0 // first color in palette
	greenIndex = 1 // next color in palette
	blueIndex  = 2
	redIndex   = 3
)

func main() {
	// The sequence of images is deterministic unless we seed
	// the pseudo-random number generator using the current time.
	// Thanks to Randall McPherson for pointing out the omission.
	rand.Seed(time.Now().UTC().UnixNano())
	lissajous(os.Stdout)
}

func lissajous(out io.Writer) {
	const (
		cycles  = 5     // number of complete x oscillator revolutions
		res     = 0.001 // angular resolution
		size    = 100   // image canvas covers [-size..+size]
		nframes = 64    // number of animation frames
		delay   = 8     // delay between frames in 10ms units
	)

	freq := rand.Float64() * 3.0 // relative frequency of y oscillator
	anim := gif.GIF{LoopCount: nframes}
	phase := 0.0 // phase difference
	colorIndex := 1
	for i := 0; i < nframes; i++ {
		rect := image.Rect(0, 0, 2*size+1, 2*size+1)
		img := image.NewPaletted(rect, palette)
		for t := 0.0; t < cycles*2*math.Pi; t += res {
			x := math.Sin(t)
			y := math.Sin(t*freq + phase)
			img.SetColorIndex(size+int(x*size+0.5), size+int(y*size+0.5),
				uint8(colorIndex))
		}
		colorIndex = (colorIndex % 7) + 1
		phase += 0.1
		anim.Delay = append(anim.Delay, delay)
		anim.Image = append(anim.Image, img)
	}
	gif.EncodeAll(out, &anim) // NOTE: ignoring encoding errors
}
```

**Comment by ToviHe** [comment]
```
// Lissajous generates GIF animations of random Lissajous figures.
package main

import (
	"image"
	"image/color"
	"image/gif"
	"io"
	"math"
	"math/rand"
	"os"
	"time"
)

// 练习1.5 调色板由黑色变成绿色
var palette = []color.Color{color.White, color.Black, color.RGBA{0xff, 0x00, 0x00, 0xff}, color.RGBA{0x00, 0xff, 0x00, 0xff}, color.RGBA{0x00, 0x00, 0xff, 0xff}}

// 练习1.6 调色板颜色增多
const (
	whiteIndex = 0 // first color in palette
	blackIndex = 1 // next color in palette
	redIndex   = 2
	greenIndex = 3
	blueIndex  = 4
)

func main() {
	// The sequence of images is deterministic unless we seed
	// the pseudo-random number generator using the current time.
	// Thanks to Randall McPherson for pointing out the omission.
	rand.Seed(time.Now().UTC().UnixNano())
	lissajous(os.Stdout)
}

func lissajous(out io.Writer) {
	const (
		cycles  = 5     // number of complete x oscillator revolutions
		res     = 0.001 // angular resolution
		size    = 100   // image canvas covers [-size..+size]
		nframes = 64    // number of animation frames
		delay   = 8     // delay between frames in 10ms units
	)
	colorIndex := uint8(0)
	freq := rand.Float64() * 3.0 // relative frequency of y oscillator
	anim := gif.GIF{LoopCount: nframes}
	phase := 0.0 // phase difference
	for i := 0; i < nframes; i++ {
		rect := image.Rect(0, 0, 2*size+1, 2*size+1)
		img := image.NewPaletted(rect, palette)
		for t := 0.0; t < cycles*2*math.Pi; t += res {
			x := math.Sin(t)
			y := math.Sin(t*freq + phase)
			img.SetColorIndex(size+int(x*size+0.5), size+int(y*size+0.5),
				colorIndex)
		}
		colorIndex = (colorIndex + 1) % uint8(len(palette))
		phase += 0.1
		anim.Delay = append(anim.Delay, delay)
		anim.Image = append(anim.Image, img)
	}
	gif.EncodeAll(out, &anim) // NOTE: ignoring encoding errors
}
```

**Comment by xm-xw** [comment]
直接down下来的代码在win环境，是有问题的，需要修改后使用。
可以在main函数中直接输出文件名
func main() {
	// The sequence of images is deterministic unless we seed
	// the pseudo-random number generator using the current time.
	// Thanks to Randall McPherson for pointing out the omission.
	rand.Seed(time.Now().UTC().UnixNano())
	create, _ := os.Create("out.gif")
	lissajous(create)
}

**Comment by GaytHub** [comment]
打卡，ps：这章好多没看懂

**Reply by Apahno** [reply]
AI时代打卡，复制粘贴了代码下来，哪里不懂搜哪里，练习答案都是看评论的我深深自责

**Comment by swanGooseMan** [comment]
练习 1.5： var palette = []color.Color{color.White, color.RGBA{0, 0xff, 0, 0xff}}

**Comment by Chen-05** [comment]
// Lissajous generates GIF animations of random Lissajous figures.
package main

import (
	"image"
	"image/color"
	"image/gif"
	"io"
	"log"
	"math"
	"math/rand"
	"os"
	"path/filepath"
	"time"
)

var palette = []color.Color{
	color.White,
	color.Black,
	color.RGBA{R: 255, A: 255},         // 红
	color.RGBA{R: 255, G: 127, A: 255}, // 橙
	color.RGBA{R: 255, G: 255, A: 255}, // 黄
	color.RGBA{G: 255, A: 255},         // 绿
	color.RGBA{B: 255, A: 255},         // 蓝
	color.RGBA{R: 75, B: 130, A: 255},  // 靛
	color.RGBA{R: 148, B: 211, A: 255}, // 紫
}

const (
	whiteIndex  = 0 // first color in palette
	blackIndex  = 1 // next color in palette
	redIndex    = 2
	orangeIndex = 3
	yellowIndex = 4
	greenIndex  = 5
	blueIndex   = 6
	indigoIndex = 7
	purpleIndex = 8
)

func main() {
	// The sequence of images is deterministic unless we seed
	// the pseudo-random number generator using the current time.
	// Thanks to Randall McPherson for pointing out the omission.
	rand.Seed(time.Now().UTC().UnixNano())

	exePath, _ := os.Getwd()
	f, err := os.Create(filepath.Join(exePath, "out.gif"))
	if err != nil {
		log.Fatalf("Create File err: %v", err)
	}
	defer func() {
		if err := f.Close(); err != nil {
			log.Printf("Close File err: %v", err)
		}
	}()

	if err := lissajous(f); err != nil {
		log.Fatalf("Create Gif err: %v", err)
	}
	log.Printf("Create gif success!\n%s\\out.gif", exePath)
}

func lissajous(out io.Writer) error {
	const (
		cycles  = 5     // number of complete x oscillator revolutions
		res     = 0.001 // angular resolution
		size    = 100   // image canvas covers [-size..+size]
		nframes = 65    // number of animation frames
		delay   = 8     // delay between frames in 10ms units
	)

	freq := rand.Float64() * 2.0 // relative frequency of y oscillator
	anim := gif.GIF{LoopCount: nframes}
	phase := 0.0 // phase difference
	for i := 0; i < nframes; i++ {
		rect := image.Rect(0, 0, 2*size+1, 2*size+1)
		img := image.NewPaletted(rect, palette)
		for t := 0.0; t < cycles*2*math.Pi; t += res {
			x := math.Sin(t)
			y := math.Sin(t*freq + phase)
			img.SetColorIndex(size+int(x*size+0.5), size+int(y*size+0.5),
				orangeIndex)
		}
		phase += 0.1
		anim.Delay = append(anim.Delay, delay)
		anim.Image = append(anim.Image, img)
	}
	return gif.EncodeAll(out, &anim) // NOTE: ignoring encoding errors
}

**Comment by lingdong0128-sys** [comment]
package main

import (
	"image"
	"image/color"
	"image/gif"
	"io"
	"math"
	"math/rand/v2"
	"os"
)

//在go中，常量在定义后在程序运行中就不会变了，常量通常是一个包级变量，除非他在函数内部
//在这个程序中，我们需要用go来制作一个gif动图，那么我们就需要帧数，圈数，延迟
//变量声明一般也是包级

// 我们先定义需要用到的颜色
// palette定义了只能使用的颜色
var palette = []color.Color{
	color.White,                              // 0号：白色 (通常作为背景底色保留)
	color.Black,                              // 1号：黑色
	color.RGBA{R: 255, G: 0, B: 0, A: 255},   // 2号：红 (Red)
	color.RGBA{R: 255, G: 165, B: 0, A: 255}, // 3号：橙 (Orange)
	color.RGBA{R: 255, G: 255, B: 0, A: 255}, // 4号：黄 (Yellow)
	color.RGBA{R: 0, G: 255, B: 0, A: 255},   // 5号：绿 (Green)
	color.RGBA{R: 0, G: 127, B: 255, A: 255}, // 6号：青 (Cyan/Light Blue)
	color.RGBA{R: 0, G: 0, B: 255, A: 255},   // 7号：蓝 (Blue)
	color.RGBA{R: 139, G: 0, B: 255, A: 255}, // 8号：紫 (Purple)
}

// 定义常量
// 定义白和黑的下标
// const(常量)
const (
	whiteindex  = 0
	blackindex  = 1
	redindex    = 2
	orangeindex = 3
	yellowindex = 4
	greenindex  = 5
	cyanindex   = 6
	blueindex   = 7
	purpleindex = 8
)

// 主函数
func main() {
	//go1.20以后rand在底层会自动给一个种子，所以已经不需要当前毫秒的时间戳作为种子了
	lissajous(os.Stdout)
}

// 画图函数
// out io.Write接受的是一个输出点，用于定位main中传递的输出点
func lissajous(out io.Writer) {
	const (
		//需要绘制5个圆周期
		cycles = 5
		//resolution分辨率
		res  = 0.001
		size = 100
		//frames(帧数)
		nframes = 64
		//delay(延迟)
		delay = 8
	)
	//频率，生成浮点数64位[0.0,1.0)*3.0的随机频率
	freq := rand.Float64() * 3.0
	//GIF是一个结构体，struct,loopcount定义了gif的循环次数
	anim := gif.GIF{LoopCount: nframes}
	//phase（相位）决定了函数的起跑点
	phase := 0.0
	//外层循环，为gif添加图片和速度属性，循环帧数次
	for i := 0; i < nframes; i++ {
		//画布大小，传递两个坐标（x0,y0）,(x1,y1),因为中心原点会占住一个像素，所以+1才是正方形
		rect := image.Rect(0, 0, 2*size+1, 2*size+1)
		//定义一个新图片的结构体并赋值给img的结构体Paletted
		img := image.NewPaletted(rect, palette)
		//开始画图
		for t := 0.0; t < cycles*2*math.Pi; t += res {
			x := math.Sin(t)
			y := math.Sin(freq*t + phase)
			colorindex := uint8(rand.IntN(7) + 2)
			img.SetColorIndex(size+int(x*size+0.5), size+int(y*size+0.5), colorindex)
		}
		//换个起始点
		phase += 1
		anim.Delay = append(anim.Delay, delay)
		anim.Image = append(anim.Image, img)
	}
	//EncodeAll,gif专用的编码接口，用于编码动画
	gif.EncodeAll(out, &anim)
}
我做了尽可能详细的注释，并完成了作业

**Comment by Fragmentmemory** [comment]
// Lissajous generates GIF animations of random Lissajous figures.
package main

import (
    "image"
    "image/color"
    "image/gif"
    "io"
    "math"
    "math/rand"
    "os"
    "time"
)

// var palette = []color.Color{color.White, color.Black}
var palette = []color.Color{
    color.White,
    color.RGBA{0x00, 0xff, 0x00, 0xff},
    color.RGBA{0x00, 0x00, 0xff, 0xff},
    color.RGBA{0xff, 0x00, 0x00, 0xff},
}

const (
    whiteIndex = 3 // first color in palette
    blackIndex = 0 // next color in palette
)

func main() {
    // The sequence of images is deterministic unless we seed
    // the pseudo-random number generator using the current time.
    // Thanks to Randall McPherson for pointing out the omission.
    rand.Seed(time.Now().UTC().UnixNano())
    create,_:= os.Create("out.gif")
    // lissajous(os.Stdout)
    lissajous(create)
}

func lissajous(out io.Writer) {
    const (
        cycles  = 5     // number of complete x oscillator revolutions
        res     = 0.001 // angular resolution
        size    = 100   // image canvas covers [-size..+size]
        nframes = 64    // number of animation frames
        delay   = 8     // delay between frames in 10ms units
    )

    freq := rand.Float64() * 3.0 // relative frequency of y oscillator
    anim := gif.GIF{LoopCount: nframes}
    phase := 0.0 // phase difference
    var index uint8 = 0
    for i := 0; i < nframes; i++ {
        rect := image.Rect(0, 0, 2*size+1, 2*size+1)
        img := image.NewPaletted(rect, palette)
        for t := 0.0; t < cycles*2*math.Pi; t += res {
            x := math.Sin(t)
            y := math.Sin(t*freq + phase)
            img.SetColorIndex(size+int(x*size+0.5), size+int(y*size+0.5),
                index)
            }
        index=(index+1)%3 +1
        phase += 0.1
        anim.Delay = append(anim.Delay, delay)
        anim.Image = append(anim.Image, img)
    }
    gif.EncodeAll(out, &anim) // NOTE: ignoring encoding errors
}
